cpp/src/barretenberg/stdlib/hash/keccak/keccak.cpp

```diff
@@ -604,12 +635,119 @@ template <typename Composer> stdlib::byte_array<Composer> keccak<Composer>::hash
         } else {
             sliced = field_ct(block_bytes.slice(i * WORD_SIZE, WORD_SIZE));
         }
-        const auto accumulators = plookup_read::get_lookup_accumulators(KECCAK_FORMAT_INPUT, sliced);
+        sliced_buffer.emplace_back(sliced);
+    }
+
+    // If the input preimage size is known at circuit-compile time, nothing more to do.
+    if (num_bytes.is_constant()) {
+        return sliced_buffer;
+    }
+
+    // If we do *not* know the preimage size at circuit-compile time, we have several steps we must execute:
+    // 1. Validate that `input[num_bytes], input[num_bytes + 1], ..., input[input.size() - 1]` are all ZERO.
```

Comment:
Why do we need to do this?

Reply:
the keccak algorithm takes the input message and concatenates with the byte 0x1. This string is then zero-padded to fill a multiple of the keccak block length. The final byte of this padded string is set to 0x80.

We have a need to compute keccak hashes in a circuit where the length of the preimage is not known, only the maximum length. This happens when computing keccak for Ethereum storage proofs (the hash preimage is the RLP-encoded node data. The RLP encoding does not produce a fixed length)

When the input is variable-length (up to a max), we compute a zero-padded byte array and dynamically insert the two formatting bytes depending on the witness value of the length parameter